src/shell/shell_misc.cpp

```diff
@@ -20,479 +20,421 @@
+			constexpr uint8_t Left         = 0x4B;
+			constexpr uint8_t Right        = 0x4D;
+			constexpr uint8_t End          = 0x4F;
+			constexpr uint8_t Down         = 0x50;
+			constexpr uint8_t Delete       = 0x53;
+			switch (data) {
+			case F3: {
+				if (history.empty()) {
 					break;
+				}
 
-				case 0x4D: /* Right */
-					if (str_index < str_len) {
-						outc(line[str_index++]);
-					}
+				std::string_view last_command = history.back();
+				if (last_command.size() <= command.size()) {
 					break;
+				}
 
-				case 0x47: /* Home */
-					while (str_index) {
-						move_cursor_back_one();
-						str_index--;
-					}
-					break;
+				last_command = last_command.substr(command.size());
+				command += last_command;
+				cursor_position = command.size();
+				break;
+			}
 
-				case 0x4F: /* End */
-					while (str_index < str_len) {
-						outc(line[str_index++]);
-					}
-					break;
+			case Left:
+				if (cursor_position > 0) {
+					cursor_position--;
+				}
+				break;
 
-				case 0x48: /* Up */
-					if (l_history.empty() || it_history == l_history.end()) break;
-
-					// store current command in history if we are at beginning
-					if (it_history == l_history.begin() && !current_hist) {
-						current_hist=true;
-						l_history.push_front(line);
-					}
-
-					for (;str_index>0; str_index--) {
-						// removes all characters
-						move_cursor_back_one();
-						outc(' ');
-						move_cursor_back_one();
-					}
-					strcpy(line, it_history->c_str());
-					len = (uint16_t)it_history->length();
-					str_len = str_index = len;
-					size = CMD_MAXLINE - str_index - 2;
-					DOS_WriteFile(STDOUT, (uint8_t *)line, &len);
-					++it_history;
-					break;
+			case Right:
+				if (cursor_position < command.size()) {
+					cursor_position++;
+				}
+				break;
 
-				case 0x50: /* Down */
-					if (l_history.empty() || it_history == l_history.begin()) break;
-
-					// not very nice but works ..
-					--it_history;
-					if (it_history == l_history.begin()) {
-						// no previous commands in history
-						++it_history;
-
-						// remove current command from history
-						if (current_hist) {
-							current_hist = false;
-							l_history.pop_front();
-						}
-						break;
-					} else {
-						--it_history;
-					}
-
-					for (; str_index > 0; str_index--) {
-						// removes all characters
-						move_cursor_back_one();
-						outc(' ');
-						move_cursor_back_one();
-					}
-					strcpy(line, it_history->c_str());
-					len = (uint16_t)it_history->length();
-					str_len = str_index = len;
-					size = CMD_MAXLINE - str_index - 2;
-					DOS_WriteFile(STDOUT, (uint8_t *)line, &len);
-					++it_history;
+			case Up:
+				if (history_index > 0) {
+					history_clone[history_index] = command;
+					history_index--;
+					command = history_clone[history_index];
+					cursor_position = command.size();
+				}
+				break;
 
+			case Down:
+				if (history_index < history_clone.size() - 1) {
```

Comment:
If `history_clone.size()` is empty, then subtracting 1 from the `size_t` will wrap it to the max value, in which case the `if` check would succeed.

Maybe need a `history_clone.empty()` pre-check?